### ISSUE TITLE
OpenSSL 3.1 unsupported method

### ISSUE THREAD

**Issue by jau87343** [FreeTDS/freetds#513]
Is TLS connections using OpenSSL 3.1 (linux) supposed to work towards MSSQL?
I get some strange connection errors
error:10000079:BIO routines::unsupported method

**Comment by freddy77** [comment]
It should work. Which distro and version are you using?

**Comment by jau87343** [comment]
Im using custom linux x64
Freetds 1.3.18
It kinda look like the connection succeeds, but sending a query causes this error message.
The only change done was adding encryption = required in the global config.

Looking at the code, there are 2 custom BIO objects used; one for auth and one for connection?
The latter do not add function pointer for ctrl* which I suspect is the cause of the error message.
Perhaps openssl 3.1 require this while previous versions didnt? Just guessing..

**Comment by freddy77** [comment]
Can you understand the line that gives the error? Can you enable logging and see the output https://www.freetds.org/userguide/logging.html ? For normal logins FreeTDS uses the same BIO to send the login so it's used also if you don't specify `encryption = required` (for shorter time however).

You could try to set the control function to the same callback. It sounds a bit weird however, it would be quite a regression.

**Comment by jau87343** [comment]
adding the same control-function in src/tds/tls.c actually solved the problem :)
(I verified that connection was indeed encrypted)

A SELECT query triggered the error while a previous EXEC query did not, so I guess it depends on returned data.

The related parts of freetds.log

dblib.c:1377:dbcmd(0x7ff2b80188e0, SELECT OkoServiceId FROM vOkoService where Overforer=1 and AvsAdresseID=24)
dblib.c:1383:dbcmd() bufsz = 38
dblib.c:6083:dbfreebuf(0x7ff2b80188e0)
dblib.c:7072:dbsqlsend(0x7ff2b80188e0)
mem.c:656:tds_free_all_results()
util.c:179:Changed query state from IDLE to WRITING
packet.c:852:Sending packet
0000 01 01 00 b2 00 00 01 00-16 00 00 00 12 00 00 00 |........ ........|
0010 02 00 00 00 00 00 00 00-00 00 01 00 00 00 53 00 |........ ......S.|
0020 45 00 4c 00 45 00 43 00-54 00 20 00 4f 00 6b 00 |E.L.E.C. T. .O.k.|
0030 6f 00 53 00 65 00 72 00-76 00 69 00 63 00 65 00 |o.S.e.r. v.i.c.e.|
0040 49 00 64 00 20 00 46 00-52 00 4f 00 4d 00 20 00 |I.d. .F. R.O.M. .|
0050 76 00 4f 00 6b 00 6f 00-53 00 65 00 72 00 76 00 |v.O.k.o. S.e.r.v.|
0060 69 00 63 00 65 00 20 00-77 00 68 00 65 00 72 00 |i.c.e. . w.h.e.r.|
0070 65 00 20 00 4f 00 76 00-65 00 72 00 66 00 6f 00 |e. .O.v. e.r.f.o.|
0080 72 00 65 00 72 00 3d 00-31 00 20 00 61 00 6e 00 |r.e.r.=. 1. .a.n.|
0090 64 00 20 00 41 00 76 00-73 00 41 00 64 00 72 00 |d. .A.v. s.A.d.r.|
00a0 65 00 73 00 73 00 65 00-49 00 44 00 3d 00 32 00 |e.s.s.e. I.D.=.2.|
00b0 34 00                  -                        |4.|

tls.c:183:in tds_push_func
util.c:179:Changed query state from WRITING to PENDING
dblib.c:4841:dbsqlok(0x7ff2b80188e0)
dblib.c:4870:dbsqlok() not done, calling tds_process_tokens()
token.c:555:tds_process_tokens(0x7ff2b8017680, 0x7ff2c0754408, 0x7ff2c075440c, 0x6914)
util.c:179:Changed query state from PENDING to READING
tls.c:160:in tds_pull_func
tls.c:160:in tds_pull_func
packet.c:410:Received packet
0000 04 01 00 38 00 71 01 00-81 01 00 00 00 00 00 08 |...8.q.. ........|
0010 00 38 0c 4f 00 6b 00 6f-00 53 00 65 00 72 00 76 |.8.O.k.o .S.e.r.v|
0020 00 69 00 63 00 65 00 49-00 64 00 fd 10 00 c1 00 |.i.c.e.I .d......|
0030 00 00 00 00 00 00 00 00-                        |........|

token.c:570:processing result tokens.  marker is  81(TDS7_RESULT)
token.c:1578:processing TDS7 result metadata.
mem.c:656:tds_free_all_results()
token.c:1603:set current_results (1 column) to tds->res_info
token.c:1610:setting up 1 columns
token.c:1547:tds7_get_data_info:
        colname = OkoServiceId
        type = 56 (int)
        server's type = 56 (int)
        column_varint_size = 0
        column_size = 4 (4 on server)
token.c:1619: name                 size/wsize      type/wtype      utype
token.c:1620: -------------------- --------------- --------------- -------
token.c:1625: OkoServiceId               4/4            56/56            0
util.c:179:Changed query state from READING to PENDING
dblib.c:4896:dbsqlok() found result token
dblib.c:1708:dbresults(0x7ff2b80188e0)
dblib.c:1713:dbresults: dbresults_state is 1 (_DB_RES_RESULTSET_EMPTY)
token.c:555:tds_process_tokens(0x7ff2b8017680, 0x7ff2c0754408, 0x7ff2c075440c, 0x6914)
util.c:179:Changed query state from PENDING to READING
token.c:570:processing result tokens.  marker is  fd(DONE)
token.c:2100:tds_process_end: more_results = 0
                was_cancelled = 0
                error = 0
                done_count_valid = 1
token.c:2119:                rows_affected = 0
token.c:2122:tds_process_end() state set to TDS_IDLE
util.c:179:Changed query state from READING to IDLE
util.c:83:logic error: cannot change query state from IDLE to PENDING
util.c:179:Changed query state from IDLE to IDLE
dblib.c:1734:dbresults() tds_process_tokens returned 0 (TDS_SUCCESS),
                        result_type 4052 (TDS_DONE_RESULT)
dblib.c:1761:dbresults(): dbresults_state is 1 (_DB_RES_RESULTSET_EMPTY)
dblib.c:1698:dbresults returning 1 (SUCCEED)
dblib.c:1850:dbnumcols(0x7ff2b80188e0)
dblib.c:2974:dbcoltype(0x7ff2b80188e0, 1)
dblib.c:3173:dbcollen(0x7ff2b80188e0, 1)
dblib.c:6470:dbprtype(56)
dblib.c:2876:dbwillconvert(SYBINT4, SYBCHAR)
convert.c:3111:tds_willconvert(56, 47)

**Comment by freddy77** [comment]
There was a bug in OpenSSL. See https://github.com/openssl/openssl/issues/18176 and https://github.com/openssl/openssl/pull/18178. Can you try an updated OpenSSL version? Like OpenSSL 3.1.4? I tried FreeTDS with a Fedora 39 which has OpenSSL 3.1.1 and it worked correctly.

**Comment by jau87343** [comment]
That bug was patched may 2022, but our openssl version 3.1 was released march 2023.
There might be another bug instead between those versions. I'll update and re-test with 3.1.4 shortly.

**Comment by jau87343** [comment]
I get the same error with openssl 3.1.4, and it still works with the Freetds ctrl patch (freetds 1.4 + openssl 3.1.4)
KTLS is enabled in openssl (threads, enable-ktls, no-zlib, shared). Kernel 5.10.46. I also did a reboot to kernel 6.1.55, but same error.

**Comment by jau87343** [comment]
More info:
Tsql CLI does not seem to fail on the same select query.
The application using the freetds ABI is using other tls-sockets  (with ktls) in other threads at the same time.
Strace of both tsql and application shows that ktls is enabled since data in recvfrom is already decrypted

I'm guessing tsql cli is using the same freetds ABI library internally, so this strange ctrl-requirement error is openssl specific

**Comment by jau87343** [comment]
Turns out this was a bit more complex, but I believe I found the root cause:

Openssl does indeed return an error because control-function is not set.
This error message is appended to Openssl's errorstack. FreeTds do however not seem to check/handle this error and continues anway. This is why some applications might work regardless of this.
The code in the application I am using do however see the openssl error and throws/exits.
